sklearn/cluster/_k_means_lloyd.pyx

```diff
@@ -0,0 +1,388 @@
+
+
+cpdef void _lloyd_iter_chunked_dense(np.ndarray[floating, ndim=2, mode='c'] X,
+                                     floating[::1] sample_weight,
+                                     floating[::1] x_squared_norms,
+                                     floating[:, ::1] centers_old,
+                                     floating[:, ::1] centers_new,
+                                     floating[::1] centers_squared_norms,
+                                     floating[::1] weight_in_clusters,
+                                     int[::1] labels,
+                                     floating[::1] center_shift,
+                                     int n_jobs,
+                                     bint update_centers=True):
+    """Single iteration of K-means lloyd algorithm with dense input.
+
+    Update labels and centers (inplace), for one iteration, distributed
+    over data chunks.
+
+    Parameters
+    ----------
+    X : {float32, float64} array-like, shape (n_samples, n_features)
+        The observations to cluster.
+
+    sample_weight : {float32, float64} array-like, shape (n_samples,)
+        The weights for each observation in X.
+
+    x_squared_norms : {float32, float64} array-like, shape (n_samples,)
+        Squared L2 norm of X.
+
+    centers_old : {float32, float64} array-like, shape (n_clusters, n_features)
+        Centers before previous iteration, placeholder for the centers after
+        previous iteration.
+
+    centers_new : {float32, float64} array-like, shape (n_clusters, n_features)
+        Centers after previous iteration, placeholder for the new centers
+        computed during this iteration.
+
+    centers_squared_norms : {float32, float64} array-like, shape (n_clusters,)
+        Squared L2 norm of the centers.
+
+    weight_in_clusters : {float32, float64} array-like, shape (n_clusters,)
+        Placeholder for the sums of the weights of every observation assigned
+        to each center.
+
+    labels : int array-like, shape (n_samples,)
+        labels assignment.
+
+    center_shift : {float32, float64} array-like, shape (n_clusters,)
+        Distance between old and new centers.
+
+    n_jobs : int
+        The number of threads to be used by openmp. If -1, openmp will use as
+        many as possible.
+
+    update_centers : bool
+        - If True, the labels and the new centers will be computed, i.e. runs
+          the E-step and the M-step of the algorithm.
+        - If False, only the labels will be computed, i.e runs the E-step of
+          the algorithm.
+    """
+    cdef:
+        int n_samples = X.shape[0]
+        int n_features = X.shape[1]
+        int n_clusters = centers_new.shape[0]
+
+        # hard-coded number of samples per chunk. Appeared to be close to
+        # optimal in all situations.
+        int n_samples_chunk = 256 if n_samples > 256 else n_samples
+        int n_chunks = n_samples // n_samples_chunk
+        int n_samples_r = n_samples % n_samples_chunk
```

Comment:
maybe use `n_samples_rem(aining)` if possible, or just add a comment about what this is